### DISCUSSION TITLE
widget.clone() vs Rc::clone(&widget)

### DISCUSSION THREAD

**Discussion by mvasi90** [fltk-rs/fltk-rs#829] [Q&A] [answered]
Hi,

I'm trying to use the same widget in functions and at the same time in closures:
```rust
...
fn main() {
...
   let mut frame = Frame::default();
...
   window.set_callback(move |_| {
      if fltk::app::event_key() == fltk::enums::Key::Escape {
         test_function(&mut frame);
      }
   });
   
   vpack.handle(move |_,e| {
      match e {
         Event::Push => {
            frame.set_label(...);
         }
      }
   });
...
}
fn test_function(frame: &mut Frame) {
   ...
}
...
```
Note: This is repeated many times. There are many closures moving the frame. But I omit it for simplicity.

This can be done using reference counting `Rc`and `RefCell` for mutability: `Rc<RefCell>`:
```rust
fn main() {
   let frame = Rc::from(RefCell::from(Frame::default()));
   ...
   let frame_rc1 = Rc::clone(&frame);

   window.set_callback(move |_| {
      if fltk::app::event_key() == fltk::enums::Key::Escape {
         test_function(&mut frame_rc1.borrow_mut());
      }
   });
```
}

But I saw your example [counter.rs](https://github.com/fltk-rs/fltk-rs/blob/master/fltk/examples/counter.rs) where you are cloning the mutable reference:
> ```rust
> let mut frame1 = frame.clone();
>    but_inc.set_callback(move |_| {
>        let label = (frame1.label().parse::<i32>().unwrap() + 1).to_string();
>        frame1.set_label(&label);
>    });
> ```

My question is:
What is the difference between cloning the mutable reference and cloning the reference counter?

`frame.clone()` is not expensive? I'm thinking that only the smart pointer is cloned. But it can be used more than once in closures and functions?

**Comment by MoAlyousef** [comment, marked as answer]
Hello

Yes they’re practically the same. Like gtk-rs, fltk-rs widgets have interior mutability. All pointers are heap allocated (as is idiomatic with FLTK). Every read and write is locked, the inner pointer is owned by the parent widget and it is responsible for managing its lifetime, and there’s a widget tracker which tracks widget deletion and will panic in that case.
